### ISSUE TITLE
New "Cards Collection" Bock

### ISSUE THREAD

**Issue by georgeolaru** [pixelgrade/nova-blocks#105]
First, I confess that I went through a lot of back and forth to make a single issue considering that there are two components in place: the card itself and the section that contains those cards. 

For reference, we will use the following definitions:
- **Card Block:** Display related pieces of information in a flexible container visually resembling a playing card.
- **Collection**: Display a list of cards placed within a coherent layout.

We already have a lot of experience in designing and developing cards throughout our themes (eg. Vasco, Felt) and I hope we can rely on this expertise to make things more reliable.

## References
Here are some examples of **Cards** that we can use as a reference ([Figma](https://www.figma.com/file/LkOxQKLMqMpZwBJQiTmmuiYK/Plugin-Nova-Blocks?node-id=1615%3A469)):
![Card block usage and references](https://user-images.githubusercontent.com/1632775/77165340-c2083300-6aba-11ea-8888-ce447862dec9.png)

And here are a few examples of the **Cards Collection** block:
![Card block usage and references](https://user-images.githubusercontent.com/1632775/77165505-1ca18f00-6abb-11ea-85c5-3e43d75b0b21.png)
 
## Guidelines
The basic idea is to prompt the user to insert a **Cards Collection** block within which he can control the number of cards available using a similar interface to the Buttons block (right side "+" sign).

![image](https://user-images.githubusercontent.com/1632775/77166629-3f34a780-6abd-11ea-8a67-5086f17e4ff8.png)

Once the number of cards are set in place (from 2 to 4), the next would be to **pick up the elements available** for each card. To make use of the rich data and adjust the layout independently, I suggest to provide a set of standard fields instead of giving the user free hand to build his cards from scratch (eg. insert an image, a heading and a paragraph). 

For this I envisioned a two-list interface where the "active elements" are grouped together and using on/off toggles you can add more or remove. I did some iterations and I choose the last one as being the closest to the current editor interface (we might come back to this decision if we find that other options are more suitable for this use-case):

![image](https://user-images.githubusercontent.com/1632775/77167093-03e6a880-6abe-11ea-93aa-2803fed393b9.png)

Next there would be some option to align the elements horizontally and to size the media image:

![image](https://user-images.githubusercontent.com/1632775/77168432-7f495980-6ac0-11ea-8f4e-e8b996c24f39.png)

For the overall block section, it would be great if we can reuse the same **Emphasis** and **Spacing** components and underlying structure from the Media Card — it would provide excellent flexibility and consistency throughout the system. In the future we might explore more styles, but for now, that would be more than enough.

![image](https://user-images.githubusercontent.com/1632775/77168481-95571a00-6ac0-11ea-8bc3-14d0ce33d4f7.png)

I'm not sure about the actual **relation** between the parent block (collection) and its siblings (cards) but I will throw this scheme here to represent my thinking throughout the design process and why I chose **a master control** at the parent block level:

![image](https://user-images.githubusercontent.com/1632775/77168214-1a8dff00-6ac0-11ea-8b77-82aef02259ad.png)

## Block Options

As a wrap-up, below is the Collection Block with all its initial options:
![image](https://user-images.githubusercontent.com/1632775/77167035-eca7bb00-6abd-11ea-854a-c938d11b4e16.png)

### [Figma Document](https://www.figma.com/file/LkOxQKLMqMpZwBJQiTmmuiYK/Plugin-Nova-Blocks?node-id=1597%3A2306)

---

@razwan @madalingorbanescu  I totally trust you to decide upon the arhitecture and structure of this block and to notice the flaws and opportunities in my thinking above. If you think it's useful to have a video-call and discuss more before making a prototype, let's do it. 💪

**Comment by razwan** [comment]
@georgeolaru while working on these blocks a few questions arose to me:

1. Shouldn't we also have controls for the heading levels of the Collection Title and Subtitle?
2. Shouldn't we also have a control for the heading level of the Card Subtitle?
3. How would a user control the display style of the meta info? 
4. Will the description field be plain text field or a rich text editor?
5. Will the Card Collection and the Card blocks share the same icon?

I guess that's it for now. 😁

**Comment by georgeolaru** [comment]
Good questions @razwan! Thanks for the follow-up and I hope that I can add some useful info below.

**1 + 2 Heading Levels controls for Collection and Card blocks**
Setting the heading level for any element it's the kind of option that I would prefer to get rid of, rather than extend it to other elements: it doesn't give any intent to the user and seems to be a technical / SEO decision. 

My suggestion is to move the "Title Level" option to the "Block Area" section, make only the "H2, H3" available and automatically lower the inner headings based on the option selected:
- Block Area Title: H2 or H3
- Block Area Subtitle: Lead Paragraph (.intro) as it doesn't need to be a heading
- Card Title: {Block Area Title + 1}
- Card Subtitle:  {Card Title + 1}

**3. How would a user control the display style of the meta info?**
The meta info will be styled by the option from Customizer (Theme Options > Fonts > Meta). If that option is not available, by default, it could be a simple div or paragraph.

**4. Will the description field be a plain text field or a rich text editor?**
I think we could start with a simple text editor that allows only basic formatting (bold, italic, links). Is that a "rich editor" actually? 😅

**5. Will the Card Collection and the Card blocks share the same icon?**
Yeah – it shouldn't be a problem as the Card block will not be allowed outside the Collection block. 

––
Hope it helps.

**Comment by razwan** [comment]
Hey @georgeolaru. I've come to a pretty solid prototype for this set of blocks. 

When you have the time could you please take a look and let me know how you feel about it and if you have any guidance for further development? The branches you should test this on are [novablocks/add/cards-collection-blocks](https://github.com/pixelgrade/nova-blocks/tree/add/cards-collection-block) and [rosa2/novablocks-cards](https://github.com/pixelgrade/rosa2/tree/novablocks-cards)

Thanks!

**Comment by georgeolaru** [comment]
That's great @razwan! It looks and works wonderfully — I really like how it turned out. It seems that the system created provides a joyful experience and I loved to play with it.  Good job! 👏

I tried to replicate the mockups from the initial post and, while architecturally everything seems fine, I discovered a few details that could be improved:

![image](https://user-images.githubusercontent.com/1632775/78129370-2b276900-7420-11ea-95c7-e28370057696.png)
![image](https://user-images.githubusercontent.com/1632775/78129387-337fa400-7420-11ea-930a-e8d8603afedc.png)


- Considering that the _"Title Level"_ option is referring to the main Collection block, I suggest to move it to the "Block Area" section.
- It would be great if we can center vertically the _"Add Card"_ button and lower its footprint (eg. halving the `margin-left` value)
- I noticed that when the _Collection Title_ and _Subtitle_ are `disabled` they are still shown in the frontend: https://share.getcloudapp.com/12u1Y2n4 It might happen with other fields too as I noticed the _Subtitle_ and _Meta_ on the fourth card (see screenshot above)
- I noticed that the _Meta_ field is placed at the bottom of the card – I think a better location would be above the _Title_. Also, there seems to be no styling applied to it.
- The default **spacing** between elements seems to be too large even it seems to be using the _small-spacing_ variable within the theme. Until we introduce an option for spacing (it could be called "Density"?), I suggest exploring another variable that it's around 10-12px in absolute size.
- The _Image container height_ lower range seems to not be able to accommodate smaller image sizes and I think we should aim it to as low as ~40px
- When the _Shrink to fit (no crop)_ option is `checked`, I noticed a background inside the container – I think we should remove any background as it would conflict with transparent images also (eg. icons)
- I did also get a warning on my installation: `Warning: Use of undefined constant Math - assumed 'Math' (this will throw an Error in a future version of PHP) in /Users/georgeolaru/Local Sites/nova/app/public/wp-content/plugins/nova-blocks/src/blocks/cards-collection/init.php on line 84`

---

That's it for now! Once the remaining of the issues are done (eg. responsive) I think we have an initial beta version ready for test and release.
